### ISSUE TITLE
cmd/link: with MacOS Mojave 10.14 beta, warning: URGENT: building for OSX, but linking against dylib during make.bash

### ISSUE THREAD

**Issue by Cospotato** [golang/go#27095]
Please answer these questions before submitting your issue. Thanks!

### What version of Go are you using (`go version`)?
`go version go1.11rc1 darwin/amd64`

### Does this issue reproduce with the latest release?
no

### What operating system and processor architecture are you using (`go env`)?
```
GOARCH="amd64"
GOBIN="/Users/CosPotato/go/bin"
GOCACHE="/Users/CosPotato/Library/Caches/go-build"
GOEXE=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="darwin"
GOOS="darwin"
GOPATH="/Users/CosPotato/go"
GOPROXY=""
GORACE=""
GOROOT="/usr/local/go"
GOTMPDIR=""
GOTOOLDIR="/usr/local/go/pkg/tool/darwin_amd64"
GCCGO="gccgo"
CC="gcc"
CXX="g++"
CGO_ENABLED="0"
GOMOD=""
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -fno-caret-diagnostics -Qunused-arguments -fmessage-length=0 -fdebug-prefix-map=/var/folders/2g/4x1xzstj4vdbj939flg1q7v00000gn/T/go-build865953866=/tmp/go-build -gno-record-gcc-switches -fno-common"
```

### What did you do?
`sudo GOROOT_BOOTSTRAP=/usr/local/go1_9 ./make.bash`
If possible, provide a recipe for reproducing the error.
A complete runnable program is good.
A link on play.golang.org is best.

### What did you expect to see?
0 warning(s), 0 error(s).

### What did you see instead?
```
Building Go cmd/dist using /usr/local/go1_9.
Building Go toolchain1 using /usr/local/go1_9.
Building Go bootstrap cmd/go (go_bootstrap) using Go toolchain1.
Building Go toolchain2 using go_bootstrap and Go toolchain1.
Building Go toolchain3 using go_bootstrap and Go toolchain2.
Building packages and commands for darwin/amd64.
# runtime/cgo
ld: warning: URGENT: building for OSX, but linking against dylib (/usr/lib/libpthread.dylib) built for (unknown). Note: This will be an error in the future.
ld: warning: URGENT: building for OSX, but linking against dylib (/usr/lib/libSystem.dylib) built for (unknown). Note: This will be an error in the future.
# os/signal/internal/pty
ld: warning: URGENT: building for OSX, but linking against dylib (/usr/lib/libSystem.dylib) built for (unknown). Note: This will be an error in the future.
# plugin
ld: warning: URGENT: building for OSX, but linking against dylib (/usr/lib/libSystem.dylib) built for (unknown). Note: This will be an error in the future.
# net
ld: warning: URGENT: building for OSX, but linking against dylib (/usr/lib/libSystem.dylib) built for (unknown). Note: This will be an error in the future.
# os/user
ld: warning: URGENT: building for OSX, but linking against dylib (/usr/lib/libSystem.dylib) built for (unknown). Note: This will be an error in the future.
# crypto/x509
ld: warning: URGENT: building for OSX, but linking against dylib (/usr/lib/libSystem.dylib) built for (unknown). Note: This will be an error in the future.
ld: warning: URGENT: building for OSX, but linking against dylib (/System/Library/Frameworks//Security.framework/Security) built for (unknown). Note: This will be an error in the future.
ld: warning: URGENT: building for OSX, but linking against dylib (/System/Library/Frameworks//CoreFoundation.framework/CoreFoundation) built for (unknown). Note: This will be an error in the future.
---
Installed Go for darwin/amd64 in /usr/local/go
Installed commands in /usr/local/go/bin
```

**Comment by Cospotato** [comment]
@agnivade Confirmed, it is not a GOROOT_BOOTSTRAP issue.
run `sudo GOROOT_BOOTSTRAP=/usr/local/go1_10 ./make.bash`
got
```
Building Go cmd/dist using /usr/local/go1_10.
Building Go toolchain1 using /usr/local/go1_10.
Building Go bootstrap cmd/go (go_bootstrap) using Go toolchain1.
Building Go toolchain2 using go_bootstrap and Go toolchain1.
Building Go toolchain3 using go_bootstrap and Go toolchain2.
Building packages and commands for darwin/amd64.
# runtime/cgo
ld: warning: URGENT: building for OSX, but linking against dylib (/usr/lib/libpthread.dylib) built for (unknown). Note: This will be an error in the future.
ld: warning: URGENT: building for OSX, but linking against dylib (/usr/lib/libSystem.dylib) built for (unknown). Note: This will be an error in the future.
# os/signal/internal/pty
ld: warning: URGENT: building for OSX, but linking against dylib (/usr/lib/libSystem.dylib) built for (unknown). Note: This will be an error in the future.
# plugin
ld: warning: URGENT: building for OSX, but linking against dylib (/usr/lib/libSystem.dylib) built for (unknown). Note: This will be an error in the future.
# net
ld: warning: URGENT: building for OSX, but linking against dylib (/usr/lib/libSystem.dylib) built for (unknown). Note: This will be an error in the future.
# os/user
ld: warning: URGENT: building for OSX, but linking against dylib (/usr/lib/libSystem.dylib) built for (unknown). Note: This will be an error in the future.
# crypto/x509
ld: warning: ld: warning: URGENT: building for OSX, but linking against dylib (/System/Library/Frameworks//Security.framework/Security) built for (unknown). Note: This will be an error in the future.URGENT: building for OSX, but linking against dylib (/usr/lib/libSystem.dylib) built for (unknown). Note: This will be an error in the future.
ld: warning: URGENT: building for OSX, but linking against dylib (/System/Library/Frameworks//CoreFoundation.framework/CoreFoundation) built for (unknown). Note: This will be an error in the future.

---
Installed Go for darwin/amd64 in /usr/local/go
Installed commands in /usr/local/go/bin
```

**Comment by agnivade** [comment]
@ianlan

**Comment by rasky** [comment]
I've changed the description back to mentioning that this happens with OSX beta.

@Cospotato which version of Xcode you have installed in your OSX beta? Is it the latest one?

**Comment by Cospotato** [comment]
@rasky 
I'm using XCode 9.4.1
```
$ /usr/bin/xcodebuild -version
Xcode 9.4.1
Build version 9F2000
```
I also installed XCode 10.0.0-beta. But I selected to the stable one with `xcode-select`.

**Comment by rasky** [comment]
@Cospotato does the error appear with Xcode 10 beta?

**Comment by ianlancetaylor** [comment]
CC @randall77 

What fresh hell is this?

**Comment by rasky** [comment]
MacOS and XCode are always supposed to upgrade together. In fact, you can use newer XCode on previous MacOS versions, but not viceversa. 

Ian, Keith, given the changes that went in Go 1.11 to start using libsystem, I think we should at least document to always use the latest XCode

**Comment by randall77** [comment]
Got me. Anyone have the source to OSX's ld?
I suspect there's some way they are detecting the version of libSystem and friends, and it is failing.  Hence the `(unknown)`.  Possibly it is the path.  @Cospotato, if you can try editing runtime/sys_darwin.go and replace all of the `libSystem.B.dylib` with `libSystem.dylib` (the latter is just a symlink the the former) and see if anything changes.

Another possibility is versioning.  If I do
```
$ otool -L /bin/ls
/bin/ls:
	/usr/lib/libutil.dylib (compatibility version 1.0.0, current version 1.0.0)
	/usr/lib/libncurses.5.4.dylib (compatibility version 5.4.0, current version 5.4.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1238.60.2)
$ otool -L ../bin/go
../bin/go:
	/usr/lib/libSystem.B.dylib (compatibility version 0.0.0, current version 0.0.0)
	/System/Library/Frameworks/CoreFoundation.framework/Versions/A/CoreFoundation (compatibility version 0.0.0, current version 0.0.0)
	/System/Library/Frameworks/Security.framework/Versions/A/Security (compatibility version 0.0.0, current version 0.0.0)
```
So somehow Go binaries don't have compatibility/current versions.  I don't know how to pick the right versions and/or record that in the binary, but if someone who is experiencing this bug could figure out at least the latter, we could test to see if that is it.

**Comment by randall77** [comment]
@Cospotato Just to be clear, these are only warnings, not errors?  You can still build binaries, and run them, just fine?

**Comment by randall77** [comment]
So I looked into this a bit more.  My understanding is as follows.  The linker records a path, compatibility version, and current version for each shared library a binary is linked against.  Additionally, each shared library also has a compatibility and current version.

At load time, the dynamic loader checks the version information for compatibility.

The current version (both the version recorded in the load command in the executable, and the version on the shared library itself) are ignored.

The compatibility version listed in the load command in the executable must be less than or equal to the compatibility version on the shared library itself.

I hacked our linker to require version 2.0.0 of libSystem.dylib, and those binaries will not run:
```
dyld: Library not loaded: /usr/lib/libSystem.B.dylib
  Referenced from: /private/var/folders/dx/k53rs1s93538b4x20g46cj_w0000gn/T/workdir/go/pkg/tool/darwin_amd64/go_bootstrap
  Reason: Incompatible library version: go_bootstrap requires version 2.0.0 or later, but libSystem.B.dylib provides version 1.0.0
```

We currently specify 0.0.0 for both compatibility version and current version. That seems maximally compatible.

The question remains: what happened during @Cospotato 's build?  The `(unknown)` part of the error sounds like the dynamic loader is unable to find the version (or even the architecture? or even the file itself?) of the libraries it is loading. This does not sound like a problem in Go, but in the OS installation itself. I've seen errors like this on the web that have happened for some bad Xcode beta versions.

TL;DR I don't think there's anything we need to do here, absent something I'm missing.

References:

- http://math-atlas.sourceforge.net/devel/assembly/MachORuntime.pdf
- https://github.com/realm/realm-cocoa/issues/2180

**Comment by robpike** [comment]
I just updated the command-line tools for my mac 10.13.6 installation, and am now getting the errors again when running make.bash:

```
# crypto/x509
ld: warning: text-based stub file /System/Library/Frameworks//CoreFoundation.framework/CoreFoundation.tbd and library file /System/Library/Frameworks//CoreFoundation.framework/CoreFoundation are out of sync. Falling back to library file for linking.
ld: warning: text-based stub file /System/Library/Frameworks//Security.framework/Security.tbd and library file /System/Library/Frameworks//Security.framework/Security are out of sync. Falling back to library file for linking.
```

**Comment by dmitshur** [comment]
Is this still an issue with the final release of macOS Mojave 10.14 (which came out today)? I can't reproduce it.

**Comment by joshcarter** [comment]
I'm seeing this when trying to build on a clean 10.14 install. Go version 1.11, Xcode version 10.0 build version 10A255.

**Comment by Txiaozhe** [comment]
I am get an error when I compile one c++ file:
ld: warning: ld: warning: URGENT: building for OSX, but linking against dylib (/usr/lib/libc++.dylib) built for (unknown). Note: This will be an error in the future.URGENT: building for OSX, but linking against dylib (/usr/lib/libSystem.dylib) built for (unknown). Note: This will be an error in the future.

with my c++ file depends /usr/lib/libSystem.dylib and /usr/lib/libc++.dylib, it is the same problem?

my mac: macos mojave 10.14

**Comment by robpike** [comment]
Why won't this issue stay open?

**Comment by agnivade** [comment]
You have to remove the WaitingForInfo label.

**Comment by dmitshur** [comment]
I think it's another instance of gopherbot fighting with people, but doing so only once (i.e., it didn't close the issue after Rob reopened it the second time). This is #21312.

**Comment by ianlancetaylor** [comment]
I can't tell if there is anything that needs to be done here.

Is anybody seeing a problem using Go sources from tip on Mojave?

**Comment by sitano** [comment]
we are having
```
/usr/local/Cellar/go/1.12.1/libexec/pkg/tool/darwin_amd64/link: running clang failed: exit status 1
ld: warning: ignoring file /var/folders/2g/3p3p4dn95yn1zzmwjrjjn57w0000gn/T/go-link-142075432/000002.o, file was built for unsupported file format ( 0x7F 0x45 0x4C 0x46 0x02 0x01 0x01 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 ) which is not the architecture being linked (x86_64): /var/folders/2g/3p3p4dn95yn1zzmwjrjjn57w0000gn/T/go-link-142075432/000002.o
ld: warning: ignoring file /var/folders/2g/3p3p4dn95yn1zzmwjrjjn57w0000gn/T/go-link-142075432/go.o, file was built for unsupported file format ( 0x7F 0x45 0x4C 0x46 0x02 0x01 0x01 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 ) which is not the architecture being linked (x86_64): /var/folders/2g/3p3p4dn95yn1zzmwjrjjn57w0000gn/T/go-link-142075432/go.o
Undefined symbols for architecture x86_64:
  "__cgo_topofstack", referenced from:
      __cgo_7e1b3c2abc8d_C2func_getnameinfo in 000017.o
      __cgo_7e1b3c2abc8d_Cfunc_getnameinfo in 000017.o
      __cgo_7e1b3c2abc8d_C2func_getaddrinfo in 000019.o
      __cgo_7e1b3c2abc8d_Cfunc_gai_strerror in 000019.o
      __cgo_7e1b3c2abc8d_Cfunc_getaddrinfo in 000019.o
      __cgo_7b5af5e1ac14_Cfunc_ZSTD_getErrorName in 000021.o
      __cgo_7b5af5e1ac14_Cfunc_ZSTD_isError in 000021.o
      ...
  "_main", referenced from:
     implicit entry/start for main executable
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)

make: *** [build_linux_dev] Error 2
```

Go 1.12.1, MacOS 10.14.3 (18D109)

**Comment by sagar-garg** [comment]
Please open this issue again. I am getting the same error on a C code. Mac OS Mojave 10.14.4

**Comment by randall77** [comment]
This issue is getting confusing and it's not clear what, if anything, needs to be fixed.
Please open a new issue if you're having problems. Please include

- Go version
- OSX version (e.g. 10.14 aka Mojave)
- Xcode version (I believe you can get that with `xcodebuild -version`), if it is installed

And tell us *exactly* what you did.

**Comment by biskit** [comment]
i believe i can describe this one:
```
bmax:~ bs$ go version
go version go1.12.5 darwin/amd64
```

```
bmax:~ bs$ uname -a
Darwin bmax 18.6.0 Darwin Kernel Version 18.6.0: Thu Apr 25 23:16:27 PDT 2019; root:xnu-4903.261.4~2/RELEASE_X86_64 x86_64
```

code:
```go
package main

import (
    "time"

    "github.com/gofrs/uuid"

    "github.com/jinzhu/gorm"
    _ "github.com/jinzhu/gorm/dialects/sqlite"
)

type Model struct {
    CreatedAt time.Time `json:"-"`
    UpdatedAt time.Time `json:"-"`
    DeletedAt *time.Time `json:"-" sql:"index"`
}

type Product struct {
    Model
    ID uuid.UUID
    Code  string
    Price uint
}

func main() {
    db, err := gorm.Open("sqlite3", "test.db")
    if err != nil {
        panic("failed to connect database")
    }
    defer db.Close()

    // Migrate the schema
    db.AutoMigrate(&Product{})

    uuid, err := uuid.NewV4()
    // Create
    db.Create(&Product{ID:uuid, Code: "L1212", Price: 1000})

    // Read
    var product Product
    db.First(&product, 1)                   // find product with id 1
    db.First(&product, "code = ?", "L1212") // find product with code l1212

    // Update - update product's price to 2000
    db.Model(&product).Update("Price", 2000)

    // Delete - delete product
    db.Delete(&product)
}
```

```
bmax:orm bs$ go build
# ~/test/orm
ld: warning: URGENT: building for OSX, but linking against dylib (/usr/lib/libpthread.dylib) built for (unknown). Note: This will be an error in the future.
ld: warning: URGENT: building for OSX, but linking against dylib (/usr/lib/libSystem.dylib) built for (unknown). Note: This will be an error in the future.
```

building any code that has cgo involved (in this case go-sqlite3 via gorm) will do it.

this only happens when using go modules.


however, when built under ~/go/src/github.com/mattn/go-sqlite3 with the command `'go build --tags "darwin"'` it produces a libsqlite.a file that when used to build an executable does not produce this warning.

**Comment by agnivade** [comment]
As @randall77 mentioned above, please open a new issue.

**Comment by biskit** [comment]
done
